segments/base_segment.go

```diff
@@ -0,0 +1,700 @@
+		// skip local variable
+		if unicode.IsUpper([]rune(fieldName)[0]) == false {
+			continue
+		}
+
+		field := fields.FieldByName(fieldName)
+		spec, ok := BaseSegmentCharacterFormat[fieldName]
+		if ok == false {
+			return ErrSegmentInvalidType
+		}
+
+		data := record[spec.Start : spec.Start+spec.Length]
+		if err := s.isValidType(spec, data); err != nil {
+			return err
+		}
+
+		value, err := s.parseValue(spec, data)
+		if err != nil {
+			return err
+		}
+
+		// set value
+		if value.IsValid() {
+			switch value.Interface().(type) {
+			case int, int64:
+				field.SetInt(value.Interface().(int64))
+			case string:
+				field.SetString(value.Interface().(string))
+			
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *BaseSegment) String() string {
+	var buf strings.Builder
+	specifications := s.toSpecifications(BaseSegmentCharacterFormat)
+	fields := reflect.ValueOf(s).Elem()
+
+	buf.Grow(BaseSegmentLength)
+	for _, spec := range specifications {
+		value := s.toString(spec.Field, fields.FieldByName(spec.Name))
+		buf.WriteString(value)
+	}
+
+	return buf.String()
+}
+
+func (s *BaseSegment) Validate() error {
+	fields := reflect.ValueOf(s).Elem()
+	for i := 0; i < fields.NumField(); i++ {
+		fieldName := fields.Type().Field(i).Name
+		if spec, ok := BaseSegmentCharacterFormat[fieldName]; ok == true {
+			if spec.Required == Required {
+				fieldValue := fields.FieldByName(fieldName)
+				if fieldValue.IsZero() {
+					return ErrRequired
+				}
+			}
+		}
+
+		funcName := s.validateFuncName(fieldName)
+		method := reflect.ValueOf(s).MethodByName(funcName)
+		if method.IsValid() {
+			err := method.Call(nil)[0]
+			if !err.IsNil() {
+				return err.Interface().(error)
+			}
+		}
+	}
+
+	return nil
+}
+
+// customized field validation functions
+// function name should be "Validate" + field name
+
+func (s *BaseSegment) ValidateTimeStamp() error {
+	if err := s.isTimestamp(s.TimeStamp); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *BaseSegment) ValidateIdentificationNumber() error {
+	if filled, err := s.filledString(s.IdentificationNumber); err == nil {
+		msg := fmt.Sprintf("the field filled by %s", filled)
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func (s *BaseSegment) ValidatePortfolioType() error {
+	switch s.PortfolioType {
+	case "C", "I", "M", "O", "R":
+		return nil
+	}
+	return errors.New("invalid value of portfolio type")
+}
+
+func (s *BaseSegment) ValidateDateOpened() error {
+	if err := s.isDate(s.DateOpened); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *BaseSegment) ValidateTermsDuration() error {
+	switch s.TermsDuration {
+	case "LOC", "001", "REV":
+		return nil
+	}
+	_, err := strconv.Atoi(s.TermsDuration)
+	if err != nil {
+		return errors.New("invalid value of terms duration")
+	}
+	return nil
+}
+
+func (s *BaseSegment) ValidateTermsFrequency() error {
+	switch s.TermsFrequency {
+	case "D", "P", "W", "B", "E", "M", "L", "Q", "T", "S", "Y", blankString:
```

Comment:
Can we make these `const` values with names that are referenced? It's a lot easier to read `CheckingCredit` than 22 -- https://github.com/moov-io/ach/blob/master/entryDetail.go#L110

Reply:
fixed